File: python/day02.py
with open("input-day02.txt") as f:
    lines = [l.strip() for l in f.readlines() if l.strip() != ""]

# Part 1
safe_count = 0
for report in lines:
    nums = [int(n) for n in report.split(" ")]

    # Case where len < 2?
    diff = nums[1] - nums[0]
    if abs(diff) > 3 or abs(diff) < 1:
        continue

    inc = (diff > 0)
    safe = True
    for i in range(2, len(nums)):
        diff = nums[i] - nums[i - 1]

        if (diff > 0) != inc:
            safe = False
            break

        if abs(diff) > 3 or abs(diff) < 1:
            safe = False
            break

    if safe:
        safe_count += 1

print(safe_count)

# Part 2
def safe(nums, bad):
    if len(nums) < 2:
        return True

    i = 0
    j = 1
    diff = nums[j] - nums[i]
    inc = (diff > 0)
    while j < len(nums):
        diff = nums[j] - nums[i]
        if abs(diff) > 3 or abs(diff) < 1 or (diff > 0) != inc:
            if bad:
                return False

            # Case where last num can be chopped off
            if j == len(nums) - 1:
                return True

            # Case where first num can be chopped off
            if i - 1 == 0:
                if safe(nums[i:], True):
                    return True

            return safe(nums[:j] + nums[j + 1:], True) or safe(nums[:i] + nums[j:], True)

        i += 1
        j += 1

    return True

safe_count = 0
for report in lines:
    nums = [int(n) for n in report.split(" ")]
    if safe(nums, False):
        safe_count += 1

print(safe_count)

File: src/day02.rs
use crate::utils;
use utils::{lines_to_ints, Res, Soln};

fn part1(input: &Vec<String>) -> Res<i32> {
    let nums = lines_to_ints(input)?;
    let safes = nums
        .into_iter()
        .map(|set| {
            let inc = (set[1] - set[0]) > 0;
            let is_nok = set
                .iter()
                .skip(1)
                .enumerate()
                .map(|(i, n)| n - set[i]) // enumerate still starts from 0
                .any(|diff| (diff > 0) != inc || diff == 0 || diff > 3 || diff < -3);

            if is_nok {
                0
            } else {
                1
            }
        })
        .sum();
    Ok(safes)
}

fn is_nok(nums: &Vec<i32>) -> bool {
    let inc = (nums[1] - nums[0]) > 0;
    nums.iter()
        .skip(1)
        .enumerate()
        .map(|(i, n)| n - nums[i])
        .any(|diff| (diff > 0) != inc || diff == 0 || diff > 3 || diff < -3)
}

#[derive(Debug)]
struct ReportData {
    report: Vec<i32>,
    problems: (usize, usize),
}

fn part2(input: &Vec<String>) -> Res<i32> {
    let nums = lines_to_ints(input)?;
    let n_reports = nums.len();
    let nok_nums: Vec<_> = nums.into_iter().filter(|nums| is_nok(&nums)).collect();

    let mut n_ok_reports = n_reports - nok_nums.len();

    let problem_info = nok_nums
        .into_iter()
        .filter_map(|report| {
            let diffs = report
                .iter()
                .skip(1)
                .enumerate()
                .map(|(i, n)| n - report[i])
                .collect::<Vec<_>>();

            let (pos, neg, zero) = diffs
                .iter()
                .scan((0, 0, 0), |(pos, neg, zero), &n| {
                    if n > 0 {
                        *pos += 1;
                    } else if n < 0 {
                        *neg += 1;
                    } else {
                        *zero += 1;
                    }
                    Some((*pos, *neg, *zero))
                })
                .last()
                .unwrap();

            if zero > 1 || (pos > 1 && neg > 1) {
                return None;
            }

            let is_inc = pos > neg;
            let mut problems = diffs
                .iter()
                .enumerate()
                .filter_map(|(i, &n)| {
                    if n == 0 || n.abs() > 3 || (is_inc && n < 0) || (!is_inc && n > 0) {
                        Some((i, i + 1))
                    } else {
                        None
                    }
                })
                .collect::<Vec<_>>();

            if problems.is_empty() {
                None
            } else {
                Some(ReportData {
                    report,
                    problems: problems.pop().unwrap(),
                })
            }
        })
        .collect::<Vec<_>>();

    for data in problem_info {
        let &(f, s) = &data.problems;
        let mut rep1 = data.report.clone();
        let mut rep2 = data.report.clone();
        rep1.remove(f);
        rep2.remove(s);

        if !is_nok(&rep1) || !is_nok(&rep2) {
            n_ok_reports += 1;
        }
    }

    Ok(n_ok_reports as i32)
}

pub const DAY02_SOLN: Soln = Soln {
    first: part1,
    second: part2,
};
